File: api/src/test/java/org/wcl/mfn/api/controller/tools/ContractCalculatorAPIControllerTest.java
package org.wcl.mfn.api.controller.tools;

import static org.assertj.core.api.Assertions.*;
import org.junit.jupiter.api.*;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import static org.mockito.Mockito.*;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.wcl.mfn.entities.contract.calculator.*;
import org.wcl.mfn.exceptions.validation.InvalidParameterException;
import org.wcl.mfn.service.tools.ContractCalculatorService;

import java.util.List;

@ExtendWith(MockitoExtension.class)
public class ContractCalculatorAPIControllerTest {
    private ContractCalculatorAPIController testController;

    @Mock
    private ContractCalculatorService mockService;

    @BeforeEach
    public void setupTestControllerAndMockService() {
        testController = new ContractCalculatorAPIController(mockService);
    }

    @Test
    public void whenValidRequestRemunerationSupplied_thenCorrectResultsSupplied() {
        final var requestRemuneration = new RequestedRemuneration(76, 773);

        final var mockedSuggestedContracts = List.<SuggestedContract>of(
                new SuggestedContract(1, 43, 6554653, 543543, List.of()),
                new SuggestedContract(2, 44, 6554654, 543544, List.of()),
                new SuggestedContract(3, 45, 6554655, 543545, List.of()),
                new SuggestedContract(4, 46, 6554656, 543546, List.of()),
                new SuggestedContract(5, 47, 6554657, 543547, List.of()),
                new SuggestedContract(6, 48, 6554658, 543548, List.of()));

        when(mockService.suggestedContracts(76, 773))
                .thenReturn(mockedSuggestedContracts);

        try {
            final var result = testController.calculateSuggestedContracts(requestRemuneration);
            assertThat(result.getStatusCode()).isEqualTo(HttpStatusCode.valueOf(200));
            assertThat(result.getBody()).isEqualTo(mockedSuggestedContracts);
        } catch (InvalidParameterException e) {
            fail("Error running test - " + e);
        }
    }

    @Test
    public void whenInvalidRequestRemunerationSupplied_thenExceptionThrown() {
        assertThatThrownBy(() -> {
            testController.calculateSuggestedContracts(null);
        }).isInstanceOf(InvalidParameterException.class)
                .hasMessageContaining("Request remuneration value cannot be null");
    }
}

File: api/src/test/kotlin/org/wcl/mfn/api/controller/tools/ContractCalculatorAPIControllerTest.kt
package org.wcl.mfn.api.controller.tools

import io.restassured.RestAssured
import io.restassured.http.ContentType
import jakarta.annotation.PostConstruct
import org.junit.jupiter.api.Test
import org.mockito.Mockito.*
import org.springframework.boot.autoconfigure.EnableAutoConfiguration
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.boot.test.web.server.LocalServerPort
import org.springframework.http.HttpStatus
import org.wcl.mfn.api.testutils.*
import org.wcl.mfn.entities.contract.calculator.*
import org.wcl.mfn.service.tools.ContractCalculatorService
import java.io.IOException

@SpringBootTest(
    classes = [ContractCalculatorAPIController::class],
    webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT
)
@EnableConfigurationProperties
@EnableAutoConfiguration
class ContractCalculatorAPIControllerTest {
    @LocalServerPort
    private val port = 0

    private var uri: String? = null

    @PostConstruct
    fun init() {
        uri = URIFactory.testURI(port, CONTRACT_CALCULATOR_ENDPOINT)
    }

    @MockBean
    var contractCalculatorService: ContractCalculatorService? = null

    @Test
    @Throws(IOException::class)
    fun givenValidRequestedRemuneration_whenRequestingSuggestedContracts_thenCalculatedSuggestedContractsReturned() {
        val expectedResponseJson = FileTestUtils.testResourceFileAsJson(VALID_JSON_FILE)

        val expectedData = JsonTestUtils.deserializeJson(
            expectedResponseJson,
            SuggestedContract::class.java
        ) as List<SuggestedContract>

        `when`(contractCalculatorService!!.suggestedContracts(695018, 72152864))
            .thenReturn(expectedData)

        RestAssured.with()
            .contentType(ContentType.JSON)
            .body(RequestedRemuneration(695018, 72152864))
            .`when`()
            .request("POST", uri)
            .then()
            .statusCode(HttpStatus.OK.value())
            .and()
            .body("", JsonValidatingMatcher.matchFullJsonCollection<Any>(expectedResponseJson))
    }

    companion object {
        private const val VALID_JSON_FILE = "json/tools/contract-calculator-valid.json"
        private const val CONTRACT_CALCULATOR_ENDPOINT = "/api/contract-calculator"
    }
}
